Client/Coordinators/Router/URLScanner.swift

```diff
@@ -0,0 +1,75 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/
+
+import Foundation
+
+/// A utility type for extracting information from URLs, specifically those with certain schemes and formats used by the Mozilla Firefox browser.
+///
+/// Usage:
+/// ```
+/// let url = URL(string: "firefox://abcdefg/foo/bar?param1=value1&param2=value2")!
+/// let scanner = URLScanner(url: url)
+/// if scanner.isOurScheme {
+///     print("This URL uses a Mozilla Firefox extension scheme.")
+///     print("The path components are: \(scanner.components)")
+///     print("The host is: \(scanner.host)")
+///     print("The value of 'param1' is: \(scanner.value(query: "param1") ?? "N/A")")
+/// }
+/// ```
+///
+/// - Note: This type is designed for use with URLs that conform to the expected format. Unexpected URLs may result in nil values and/or incorrect information being returned.
+struct URLScanner {
+    /// The path components of the URL, excluding the scheme, host, and query parameters.
+    var components: [String]
+
+    /// The query parameters of the URL, as an array of `URLQueryItem` objects.
+    var queries: [URLQueryItem]
+
+    /// The scheme of the URL, such as "http" or "https".
+    let scheme: String
+
+    /// The host of the URL, such as "example.com".
+    let host: String
+
+    /// Initializes a new `URLScanner` object with the specified URL, if the URL is valid and uses a recognized scheme.
+    ///
+    /// - Parameter url: The URL to scan.
+    ///
+    /// - Returns: A new `URLScanner` object if the URL is valid and uses a recognized scheme, otherwise `nil`.
+    init?(url: URL) {
+        let url = url.sanitized
```

Comment:
I think the `url.sanitized` extension should be in this `URLScanner` struct since that's the only place we use it. Also, I think we should have tests specifically related to `URLScanner`